Sources/NIOExtras/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+///
+/// A decoder that splits the received `ByteBuffer` by the number of bytes speicifed in a fixed length header
+/// contained within the buffer.
+/// For example, if you received the following four fragmented packets:
+///     +---+----+------+----+
+///     | A | BC | DEFG | HI |
+///     +---+----+------+----+
+///
+/// Given that the specified header length is 1 byte,
+/// where the first header specifies 3 bytes while the second header speicifies 4 bytes,
+/// a `LengthFieldBasedFrameDecoder` will decode them into the following packets:
+///
+///     +-----+------+
+///     | BCD | FGHI |
+///     +-----+------+
+///
+/// 'A' and 'E' will be the headers and will not be passed forward.
+///
+
+public final class LengthFieldBasedFrameDecoder: ByteToMessageDecoder {
+
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    
+    private let lengthFieldLength: Int
+    private let lengthFieldEndianness: Endianness
+    
+    /// Create `LengthFieldBasedFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: Int, lengthFieldEndianness: Endianness = .little) {
```

Comment:
This is not a great choice for the default endianness. In general there are three major byte orders: little endian, big endian, and "network byte order". "Network byte order" is so-called because it is the byte ordering used by almost all network protocols, and it is the same as big endian.

For this purpose, we should default to big endian: users should be required to opt-in to little-endian.

Reply:
Okay, will do both of those.